### PR TITLE
Update modal component branch

### PR DESCRIPTION
**Pull request by Caleb-T-Owens**

### CODE REVIEW

File: app/src/lib/branches/dragActions.ts [hunk 1]
```diff
@@ -0,0 +1,53 @@
+import { DraggableCommit, DraggableHunk, DraggableFile } from '$lib/dragging/draggables';
+import { filesToOwnership } from '$lib/vbranches/ownership';
+import type { BranchController } from '$lib/vbranches/branchController';
+import type { Branch } from '$lib/vbranches/types';
+
+class BranchDragActions {
+	constructor(
+		private branchController: BranchController,
+		private branch: Branch
+	) {}
```
Comment (Caleb-T-Owens) [inline app/src/lib/branches/dragActions.ts:10]: I was thinking about whether I should have some
```ts
interface DropzoneActions {
  accepts: //...
  onDrop: //...
}
```
So we'd have one class for each dropzone actions. It would also mean that we could pass something that implements that interface directly to the Dropzone component, but I didn't quite want to go that far yet.

File: app/src/lib/branches/dragActions.ts [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+	}
+}
+
+export class BranchDragActionsFactory {
+	constructor(private branchController: BranchController) {}
+
+	build(branch: Branch) {
+		return new BranchDragActions(this.branchController, branch);
+	}
+}
```
Comment (Caleb-T-Owens) [inline app/src/lib/branches/dragActions.ts:53]: I'm using a factory so that when we go to use the drag actions, we only need to provide the relevant state, rather than having to also provide a bunch of controllers and servicies

File: app/src/lib/components/BranchDropzone.svelte [hunk 3]
```diff
@@ -31,49 +31,45 @@
...
+<div class="canvas-dropzone">
+	<Dropzone {accepts} ondrop={onDrop}>
+		{#snippet overlay({ hovered, activated })}
+			<div class="new-virtual-branch" class:activated class:hovered>
+				<div class="new-virtual-branch__content">
+					<div class="stimg">
+						<div class="stimg__hand">
+							{@html handSvg}
+						</div>
+						<div class="stimg__top-sheet">
```
Comment (Caleb-T-Owens) [inline app/src/lib/components/BranchDropzone.svelte:43]: Because this is quite a specialised dropzone, I chose not to have some generic component for the overlay that it uses.

File: app/src/lib/components/BranchLanePopupMenu.svelte [hunk 4]
```diff
@@ -159,7 +159,7 @@
 			kind="solid"
 			on:click={async () => {
 				await branchController.deleteBranch(branch.id);
-				deleteBranchModal.close();
+				close();
```
Comment (Caleb-T-Owens) [inline app/src/lib/components/BranchLanePopupMenu.svelte:162]: 😁 a cheeky unrelated change... if only we could work in multiple branches at the same time

File: app/src/lib/components/CommitDragItem.svelte [hunk 5]
```diff
@@ -1,131 +1,60 @@
...
-		return (data: DraggableCommit) => {
-			if (data.commit.isParentOf(commit)) {
-				branchController.squashBranchCommit(data.branchId, commit.id);
-			} else if (commit.isParentOf(data.commit)) {
-				branchController.squashBranchCommit(data.branchId, data.commit.id);
-			}
-		};
-	}
+		return commitDragActionsFactory.build($branch, commit);
+	});
```
Comment (Caleb-T-Owens) [inline app/src/lib/components/CommitDragItem.svelte:24]: I was considering saying something like `$derived($branch && commitDragActionsFactory.build($branch, commit))`, but I decided that I really just want to keep everything as simple as possible, and being explicit and having that sort of guard clause to me just expressed the intent much better

File: app/src/lib/components/CommitDragItem.svelte [hunk 6]
```diff
@@ -1,131 +1,60 @@
...
-		<DropzoneOverlay class="amend-dz-marker" label="Amend" />
-		<DropzoneOverlay class="squash-dz-marker" label="Squash" />
-
-		<slot />
+	<div class="commit-card-wrapper">
+		{#if actions}
+			{@render ammendDropzone()}
+		{:else}
+			{@render children()}
+		{/if}
```
Comment (Caleb-T-Owens) [inline app/src/lib/components/CommitDragItem.svelte:33]: This component needs to gracefully fall back to not talking to the branchController because its used in the remoteBranch and trunk pages.

I thought about having the snippets inside the if block too, but it ended up looking too messy

File: app/src/lib/components/CommitDragItem.svelte [hunk 7]
```diff
@@ -1,131 +1,60 @@
...
+
+{#snippet squashDropzone()}
+	<Dropzone accepts={actions!.acceptSquash.bind(actions)} ondrop={actions!.onSquash.bind(actions)}>
+		{@render children()}
+
+		{#snippet overlay({ hovered, activated })}
+			<CardOverlay {hovered} {activated} label="Squash commit" />
+		{/snippet}
+	</Dropzone>
+{/snippet}
```
Comment (Caleb-T-Owens) [inline app/src/lib/components/CommitDragItem.svelte:56]: If anyone has n idea about how to better express the nesting, I am all ears

File: app/src/lib/components/CommitList.svelte [hunk 8]
```diff
@@ -97,8 +107,31 @@
...
+	}
 </script>
 
+{#snippet reorderDropzone(dropzone: ReorderDropzone, yOffsetPx: number)}
+	<Dropzone accepts={dropzone.accepts.bind(dropzone)} ondrop={dropzone.onDrop.bind(dropzone)}>
+		{#snippet overlay({ hovered, activated })}
+			<LineOverlay {hovered} {activated} {yOffsetPx} />
+		{/snippet}
+	</Dropzone>
+{/snippet}
```
Comment (Caleb-T-Owens) [inline app/src/lib/components/CommitList.svelte:132]: I didn't think an entire component was required for the Reordering dropzone. I'm willing to be swayed on this position though.

File: app/src/lib/dragging/dropzone.ts [hunk 9]
```diff
@@ -1,37 +1,123 @@
...
+		if (this.registered) {
+			this.unregister();
+		}
+
+		// Register listeners and keep references to the functions
+		this.registered = true;
+
+		this.registeredOnDrop = async (e: DragEvent) => await this.onDrop(e, data);
+		this.registeredOnDragEnter = this.onDragEnter.bind(this);
+		this.registeredOnDragLeave = this.onDragLeave.bind(this);
```
Comment (Caleb-T-Owens) [inline app/src/lib/dragging/dropzone.ts]: I do make the assumption that the user can't drag two things at once, but that seems like a pretty sound assumption to me

File: app/src/lib/dragging/dropzone.ts [hunk 10]
```diff
@@ -1,37 +1,123 @@
...
+	private onDragLeave(e: DragEvent) {
+		e.preventDefault();
+		if (!this.active) return;
+		this.configuration.onHoverEnd();
+	}
+}
+
+export const dropzoneRegistry = new Map<HTMLElement, Dropzone>();
+
+export function dropzone(node: HTMLElement, opts: DropzoneConfiguration) {
```
Comment (Caleb-T-Owens) [inline app/src/lib/dragging/dropzone.ts]: This is now only consumed by the Dropzone, so its not relevant to support a partial configuration that contains defaults

File: app/src/lib/dragging/reorderDropzoneManager.ts [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+export class ReorderDropzoneManagerFactory {
+	constructor(private branchController: BranchController) {}
+
+	build(branch: Branch, commits: Commit[]) {
+		return new ReorderDropzoneManager(this.branchController, branch, commits);
+	}
+}
+
+// Private classes used to calculate distances between commtis
+class Indexer {
```
Comment (Caleb-T-Owens) [inline app/src/lib/dragging/reorderDropzoneManager.ts:63]: I decided to move the indexing and distance calculations into their own classes, because its now on longer required to be publicly exported (like it was when we used the old ReorderDropzone.svelte component)

File: app/src/lib/branches/dragActions.ts (hunk 1 above)
Comment (ndom91) [reply app/src/lib/branches/dragActions.ts:10]: Yeah, makes sense. Definitely sounds like a good next step. However as you said, lets maybe save that for a more full-scale refactor of the whole drag-n-drop system :+1:

File: app/src/lib/components/CommitDragItem.svelte (hunk 5 above)
Comment (ndom91) [reply app/src/lib/components/CommitDragItem.svelte:24]: Yup, makes it more readable too, etc. etc. Good call :+1:

**Review by ndom91 — APPROVED** [review]
LGTM functionality-wise!

I have a few small niggles design-wise, but you mentioned you still need to talk to Pavel so I'll just hold off :blush:

**Comment by Caleb-T-Owens** [comment]
Waiting on https://github.com/sveltejs/svelte-eslint-parser/pull/540 in order for our linting to support nested snippets.
